java/src/main/java/ai/rapids/cudf/HostColumnVector.java

```diff
@@ -738,50 +738,126 @@ public ColumnBuilder appendLists(List... inputLists) {
     public ColumnBuilder appendStructValues(List<StructData> inputList) {
       for (StructData structInput : inputList) {
         // one row
-        rows++;
         append(structInput);
       }
       return this;
     }
 
-    private void initAndResizeOffsetBuffer(int currentIndex) {
-      if (this.offsets == null) {
-        offsets = HostMemoryBuffer.allocate((rows + 1) * OFFSET_SIZE);
-        offsets.setInt(0, 0);
-      } else {
-        if (offsets.length <= currentIndex * OFFSET_SIZE + OFFSET_SIZE) {
-          HostMemoryBuffer newOffset = HostMemoryBuffer.allocate(offsets.length * 2);
+
+    /**
+     * A method that is responsible for growing the buffers as needed
+     * and incrementing the row counts when we append values or nulls.
+     * @param hasNull indicates whether the validity buffer needs to be considered, as the
+     *                nullcount may not have been fully calculated yet
+     * @param length the data length we want to fit in the data buffer
+     */
+    private void growBuffersandRows(boolean hasNull, int length) {
```

Comment:
@revans2 , I had to pass length for the string case and the hasNull flag when nullcounts haven't been set just yet. If you think there is a better way to do this, I would appreciate the comments.

Reply:
nit: and should be capitalized in the name of the method.

It is a lot closer.  Because of strings it does make things difficult.  Perhaps we should pull off the bandaid and make strings look like a list on both the CPU and the GPU.  But not now.

I would like to see us keep track of two numbers. 

1. Number of rows we have allocated space for.
2. Number of rows we have populated.

For strings we need to do it twice. Once for the offsets/validity and another once for the data.

To me it is not 100% clear what `rows` is vs `currentIndex` vs `currentByteIndex`.

I would really like to see the algorithm look like.

1. Check if we would go over row limits if we were to add another row, and blow up if we would.
2. Increment the number of rows populated and the length for strings.
3. If the rows allocated < rows populated (include data for strings)
  a. set the new number of rows allocated (use estimated row count if this is the first time else double the current value) This value needs to be capped at the row limit so we don't allocate too much memory without need.
  b. realloc the buffers so that they can hold the new allocated sizes.

Reply:
In my latest change, allocatedRows can be derived from the size of the data buffer except strings, to avoid separate tracking.
_We dont think of growing buffers (doubling essentially) in terms of rows allocated._ 
Other than that, I think this addresses the review comments.